### ISSUE TITLE
Using in monorepo (yet another question/proposal)

### ISSUE THREAD

**Issue by b12k** [release-it/release-it#858]
## Actual question / proposal

I saw https://github.com/release-it/release-it/issues/831

But I think it is still feasible to use "original" release-it, with simple (hehehe 🤣) adjustments.

Let assume we have following monorepo structure (generated by turborepo for example):

```
/
  .git
  packages/
    library-a/
      src/
      release-it.json
      package.json
    library-b/
      src/
      release-it.json
      package.json
  package.json
```

There are 2 libraries A and B. They both have respective `package.json` files and `release-it.json` configurations. So IF they were individual repos - everything would work as described in docs (default use case). The only difference between polyrepo and monorepo in this example is git history - independent and mixed.

I thinking if it will be possible to filter commits which affect affect a respective path in the monorepo.

```bash
> git log -- packages/library-a
```

👆 will returns return list of commits which affect files in `library-a`.

So something like: 

```bash
> cd ./packages/library-a
> release-it --affected=packages/library-a
```

or 

```bash
> cd ./packages/library-a
> release-it --monorepo
```

Where `--monorepo` flag will tell `release-it` to determine folder where it was executed relative to `.git` root (traverse up the folder tree) and use this path as "commit filter".

_All this cli flags should be configurable also via config file._

I do understand that `release-it` functionality depends on `tags`. But why not just ignore them in this "mode" and just work with version in `package.json` bumping it according to the flag provided (major, minor, patch)

OR 

When used with "convention" plugins provide:

- "initial version" as a cli flag `--initial-version=0.0.0` or in config `{ "initialVersion": "0.0.0" }`)
- filtered commits for affected path 

Calculate version changes relatively to initial version AND use release commits (by matching message pattern) as release markers.

Basically, always re-calculate version from initial value provided (and re-generating CHANGELOG, for example).  

Notes:
 - This approach may not work in the monorepos which are managed using `nx` as its citizens do not have individual `package.json` files.
 - Creating github releases and tags for individual items in mono-repository makes no sense
   - ❌ No taging (can be automatically disabled in "monorepo mode")
   - ❌ No GitHub releases (can be automatically disabled in "monorepo mode")
   - ✔ **But pushing to NPM, creating CHANGELOG, pushing to separate "release" repositories still does**
   - ✔ **Using monorepo's GitHub Package registry for NPM still makes sense (storing different lib packages in same monorepo registry)**

**Comment by webpro** [comment]
Thanks for this explanation, @b12k! Interesting.

So if I understand correctly, it would execute something like this in every package dir:

```
release-it --git.changelog="git log -- packages/library-a" --no-git.tag
```

- The custom `git.changelog` probably needs work
- `npm.publish` is enabled and `github.release` disabled by default
- `git.requireCommits` is `false` by default

Are you saying the when `git.changelog` command for package dir returns no commits ("not affected") it does not require a release? Perhaps that could be combined with basically a graceful `git.requireCommits: true`.

Does this idea have versioning per package separately?

And then eventually perhaps some kind of bump and push in the root dir?

I think a "monorepo" with a single `package.json` (Nx style) could be handled as a whole by release-it, since it's a single Git repo and apparently a single npm package?

**Comment by nartc** [comment]
@webpro Just want to chime in here. I'm one of the core team members of Nx and also using Nx to develop my open source projects.

`release-it` works fine for Nx monorepos with one main caveat: `package.json` (or `package-lock.json`) cannot be used as a source of truth for bumping versions as they are implicit dependencies of the caches. If they change, the cache is invalidated. The recommended way is to bump the version of the packages in the `dist/*` directory after building the packages.

Here's my `.release-it.json` for quick reference:

```json
{
  "plugins": {
    "@release-it/conventional-changelog": {},
    "@release-it/bumper": {
      "in": "version.json", // using arbitrary json file to hold the version instead
      "out": "version.json"
    }
  },
  "hooks": {
    "before:bump": "npm run build", // run build before changes are made so the cache persists
    "after:bump": [
      "git checkout -- package.json", // check out package.json
      "git checkout -- package-lock.json", // check out package-lock.json
      "node tools/scripts/update-version.mjs" // custom script to update version in `dist/*`
    ]
  }
}
```

The `update version` part can maybe done in a custom release-it plugin I suppose.

**Comment by nartc** [comment]
@webpro That...worked. I never thought about doing it that way. Cool *deleting my script*

`package.json` and `package-lock.json` are reset because there might be other pipelines that depend on the cache after release runs (eg: build documentations on the CI. Documentation site like an Examples site use the workspace's libraries and still need the cache when build).

**Comment by webpro** [comment]
Hey guys, wanted to share this, as I think it might fit with your workflows: https://github.com/webpro/release-it-monorepo

It has [wireit](https://github.com/google/wireit) as an additional dependency and simply runs the release-it process for each `package/*`. This has the advantage that each package can have its own release-it config, and with the patch for release-it itself (will release-it soon), it shows `git.changelog` (and I guess also changelog plugins) can work with filtered git logs within the package directories (this works in the demo repo).

**Comment by h3adache** [comment]
@nartc can you share an example project with this working?

**Comment by nartc** [comment]
@h3adache 
- [Angular Three](https://github.com/nartc/angular-three) sync versions. All packages share the same version
- [Nartc Workspace](https://github.com/nartc/nartc-workspace) separate versions. Each package has their own version

**Comment by ulisesbocchio** [comment]
You can pass the `path` option to [git-raw-commits](https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/git-raw-commits#gitoptspath) for the changelog plugin to select the proper package bump.
What I'm not clear is, what role does `git.changelog` play in all this? is that command only used for the gitlab/github release notes?

```json
{
  "plugins": {
    "@release-it/conventional-changelog": {
        "gitRawCommitsOpts": {
            "path": "packages/library-a"
        }
    },
    "@release-it/bumper": {
      "in": "version.json",
      "out": ["version.json", "dist/**/package.json"]
    }
  },
  "hooks": {
    "before:bump": "npm run build",
    "after:bump": [
      "git checkout -- package.json", // check out package.json
      "git checkout -- package-lock.json", // check out package-lock.json
    ]
  }
}
```

**Comment by webpro** [comment]
> What I'm not clear is, what role does `git.changelog` play in all this? is that command only used for the gitlab/github release notes?

- The output of `git.changelog` is printed in the shell when starting the process, to help users to decide the bump (when not using a plugin such as conventional-changelog).
- It's also used as default for the gitlab/hub notes release indeed, if not overridden with e.g. `github.releaseNotes`.

It's in this monorepo discussion, because in specific workflows/setups changelogs should be made/filtered to contain only changes in subdirectories

When running release-it multiple times from/for each package, the default `git.changelog` currently would still fetch all commits. Adding ` -- .` to this `git log` command could fix this.

**Comment by yianni-ververis** [comment]
@webpro any NX templates with release-it? I tried both release-it & jscutlery/semver , publishing to github packages without success. Thnx

**Comment by webpro** [comment]
You have to be more specific, I'm not sure what you're after. Please open a new ticket, this one is closed.

**Comment by hardikkaji** [comment]
> @h3adache
> 
> * [Angular Three](https://github.com/nartc/angular-three) sync versions. All packages share the same version
> * [Nartc Workspace](https://github.com/nartc/nartc-workspace) separate versions. Each package has their own version

About Each package has their own version. How do you use it with CI (Github Actions). 

How do you release those packages today? May be manually?

**Comment by hardikkaji** [comment]
> @h3adache
> 
> * [Angular Three](https://github.com/nartc/angular-three) sync versions. All packages share the same version
> * [Nartc Workspace](https://github.com/nartc/nartc-workspace) separate versions. Each package has their own version

@nartc, About Each package has their own version. How do you use it with CI (Github Actions). 

How do you release those packages today? May be manually?
